lib/Macros/Sources/SwiftMacros/PointerBoundsMacro.swift

```diff
@@ -0,0 +1,618 @@
+        var types = argTypes
+        let param = getParam(signature, index)
+        types[index] = try transformType(param.type, variant, isSizedBy)
+        if variant.skipTrivialCount {
+            if let countVar = countExpr.as(DeclReferenceExprSyntax.self) {
+                let i = try getParameterIndexForDeclRef(signature.parameterClause.parameters, countVar)
+                types[i] = nil as TypeSyntax?
+            }
+        }
+        return try base.buildFunctionSignature(types, variant)
+    }
+
+    func buildBoundsChecks(_ variant: Variant) throws -> [CodeBlockItemSyntax.Item] {
+        var res = try base.buildBoundsChecks(variant)
+        let countName: TokenSyntax = "_\(raw: name)Count"
+        let count: VariableDeclSyntax = try VariableDeclSyntax("let \(countName): some BinaryInteger = \(countExpr)")
+        res.append(CodeBlockItemSyntax.Item(count))
+
+        let countCheck = ExprSyntax("""
+        if \(getCount(variant)) < \(countName) || \(countName) < 0 {
+            fatalError("bounds check failure when calling unsafe function")
+        }
+        """)
+        res.append(CodeBlockItemSyntax.Item(countCheck))
+        return res
+    }
+
+    func unwrapIfNullable(_ expr: ExprSyntax) -> ExprSyntax {
+        if nullable {
+            return ExprSyntax(ForceUnwrapExprSyntax(expression: expr))
+        }
+        return expr
+    }
+
+    func unwrapIfNonnullable(_ expr: ExprSyntax) -> ExprSyntax {
+        if !nullable {
+            return ExprSyntax(ForceUnwrapExprSyntax(expression: expr))
+        }
+        return expr
+    }
+
+    func castIntToTargetType(expr: ExprSyntax, type: TypeSyntax) -> ExprSyntax {
+        let idType = type.as(IdentifierTypeSyntax.self)!
+        if idType.name.text == "Int" {
+            return expr
+        }
+        return ExprSyntax("\(type)(exactly: \(expr))!")
+    }
+
+    func buildUnwrapCall(_ argOverrides: [Int: ExprSyntax], _ variant: Variant) throws -> ExprSyntax {
+        let unwrappedName = TokenSyntax("_\(name)Ptr")
+        var args = argOverrides
+        let argExpr = ExprSyntax("\(unwrappedName).baseAddress")
+        assert(args[index] == nil)
+        args[index] = unwrapIfNonnullable(argExpr)
+        let call = try base.buildFunctionCall(args, variant)
+        let ptrRef = unwrapIfNullable(ExprSyntax(DeclReferenceExprSyntax(baseName: name)))
+
+        let returnKw: String = if hasReturnType(signature) {
+            "return "
+        } else {
+            ""
+        }
+        let funcName = isSizedBy ? "withUnsafeBytes" : "withUnsafeBufferPointer"
+        let unwrappedCall = ExprSyntax("""
+        \(ptrRef).\(raw: funcName) { \(unwrappedName) in
+            \(raw: returnKw)\(call)
+        }
+        """)
+        return unwrappedCall
+    }
+
+    func getCount(_ variant: Variant) -> ExprSyntax {
+        let countName = isSizedBy && variant.generateSpan ? "byteCount" : "count"
```

Comment:
The distinction I had in mind for `Variant` was to hold things that are true for all parameters in the function call, i.e. either we generate spans or we don't, and either we remove all count parameters or none. This was more relevant when generating multiple overloads, since if you had multiple countedBy parameters you wouldn't want to generate all possible combinations of span vs UBP, and all combinations of implicit vs explicit count variable.

When only ever generating a single overload, I think it makes more sense to bake this information into the builder object for each parameter's builder object.